server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java

```diff
@@ -0,0 +1,321 @@
+        blobStoreTransferService = new BlobStoreTransferService(blobStoreRepository.blobStore(), this.threadPool);
+        pinnedTimestampsBlobStore = new RemoteStorePinnedTimestampsBlobStore(
+            blobStoreTransferService,
+            blobStoreRepository,
+            clusterName,
+            this.threadPool,
+            ThreadPool.Names.REMOTE_STATE_READ
+        );
+    }
+
+    private void startAsyncUpdateTask() {
+        asyncUpdatePinnedTimestampTask = new AsyncUpdatePinnedTimestampTask(logger, threadPool, pinnedTimestampsSchedulerInterval, true);
+    }
+
+    /**
+     * Pins a timestamp in the remote store.
+     *
+     * @param timestamp The timestamp to be pinned
+     * @param pinningEntity The entity responsible for pinning the timestamp
+     * @param listener A listener to be notified when the pinning operation completes
+     * @throws IllegalArgumentException If the timestamp is less than the current time minus one second
+     */
+    public void pinTimestamp(long timestamp, String pinningEntity, ActionListener<Void> listener) {
+        // If a caller uses current system time to pin the timestamp, following check will almost always fail.
+        // So, we allow pinning timestamp in the past upto some buffer
+        long lookbackIntervalInMills = CLUSTER_REMOTE_STORE_PINNED_TIMESTAMP_LOOKBACK_INTERVAL.get(settings).millis();
+        if (timestamp < TimeUnit.NANOSECONDS.toMillis(System.nanoTime()) - lookbackIntervalInMills) {
+            throw new IllegalArgumentException(
+                "Timestamp to be pinned is less than current timestamp - value of cluster.remote_store.pinned_timestamps.lookback_interval"
+            );
+        }
+        updatePinning(pinnedTimestamps -> pinnedTimestamps.pin(timestamp, pinningEntity), listener);
+    }
+
+    /**
+     * Unpins a timestamp from the remote store.
+     *
+     * @param timestamp The timestamp to be unpinned
+     * @param pinningEntity The entity responsible for unpinning the timestamp
+     * @param listener A listener to be notified when the unpinning operation completes
+     */
+    public void unpinTimestamp(long timestamp, String pinningEntity, ActionListener<Void> listener) {
+        updatePinning(pinnedTimestamps -> pinnedTimestamps.unpin(timestamp, pinningEntity), listener);
+    }
+
+    private void updatePinning(Consumer<PinnedTimestamps> updateConsumer, ActionListener<Void> listener) {
+        RemotePinnedTimestamps remotePinnedTimestamps = new RemotePinnedTimestamps(
+            clusterService.state().metadata().clusterUUID(),
+            blobStoreRepository.getCompressor()
+        );
+        BlobPath path = pinnedTimestampsBlobStore.getBlobPathForUpload(remotePinnedTimestamps);
+        try {
+            if (updateTimetampPinningSemaphore.tryAcquire(10, TimeUnit.MINUTES)) {
+                ActionListener<Void> semaphoreAwareListener = ActionListener.runBefore(listener, updateTimetampPinningSemaphore::release);
```

Comment:
any unhandled exception is not logged in the listener chaining, we should log it to get visibility into unexpected failures